### PR TITLE
0rtt garbage resumption

### LINKED ISSUE #422: garbage 0-RTT handshake
# New test script idea

## What TLS message this idea relates to?

ClientHello, application data

## What TLS extension this idea relates to?

`early_data`

## What is the behaviour the test script should test?

Check that server, especially if it does not support 0-RTT, is still able to handle ClientHello with early_data extension and the following application data and continue a regular handshake.

Verify that garbage records are ignored by server, and following real client Finished is accepted.

## Are there scripts that test related functionality?

none

## Additional information

[draft-28](https://tools.ietf.org/html/draft-ietf-tls-tls13-28#section-4.2.10)

### PR DESCRIPTION
**Pull request by tomato42**
### Description
test if server is 0-RTT resumption tolerant, even if it does not support 0-RTT (section 4.2.10 of draft-28)

### Motivation and Context

server can turn 0-rtt support at-will, but client will not know when server does it, so the server needs to be able to process 0-rtt handshake always

fixes #422 
blocked on https://github.com/tomato42/tlslite-ng/issues/290

### Checklist

- [X] I have read the [CONTRIBUTING.md](https://github.com/tomato42/tlsfuzzer/blob/master/CONTRIBUTING.md) document and my PR follows [change requirements](https://github.com/tomato42/tlsfuzzer/blob/master/CONTRIBUTING.md#change-requirements) therein
- [x] the changes are also reflected in documentation and code comments
- [x] all new and existing tests pass (see Travis CI results)
- [x] [test script checklist](https://github.com/tomato42/tlsfuzzer/wiki/Test-script-checklist) was followed for new scripts
- [x] new test script added to `tlslite-ng.json` and `tlslite-ng-random-subset.json`
- [x] new and modified scripts were ran against popular TLS implementations:
  - [x] OpenSSL
  - [x] NSS
  - [x] GnuTLS
- [x] required version of tlslite-ng updated in requirements.txt and README.md
- [x] 0-RTT and Change Cipher Spec? discuss if the client should send CCS before early_data

---
This change is [Reviewable](https://reviewable.io/reviews/tomato42/tlsfuzzer/423)

### CODE REVIEW

File: scripts/test-tls13-0rtt-garbage.py [hunk 1]
```diff
@@ -589,6 +589,53 @@ def main():
...
+    node = node.add_child(SetRecordVersion((3, 3)))
+    node = node.add_child(ApplicationDataGenerator(getRandomBytes(num_bytes)))
+    node = node.add_child(TCPBufferingDisable())
+    node = node.add_child(TCPBufferingFlush())
+    node = node.add_child(ExpectServerHello(version=(3, 3)))
+    node = node.add_child(ExpectCertificate())
+    node = node.add_child(ExpectServerHelloDone())
+    # section D.3 of draft 28 states that client that receives TLS 1.2
+    # ServerHello as a reply to 0-RTT Client Hello MUST fail a connection
+    # consequently, it server does not need to be able to ignore early data
```
Comment (nmav) [inline scripts/test-tls13-0rtt-garbage.py]: typo: `it server`

File: scripts/test-tls13-0rtt-garbage.py [hunk 2]
```diff
@@ -589,6 +589,53 @@ def main():
...
+    node = node.add_child(TCPBufferingEnable())
+    node = node.add_child(ClientHelloGenerator(ciphers, extensions=ext))
+    node = node.add_child(SetRecordVersion((3, 3)))
+    node = node.add_child(ApplicationDataGenerator(getRandomBytes(num_bytes)))
+    node = node.add_child(TCPBufferingDisable())
+    node = node.add_child(TCPBufferingFlush())
+    node = node.add_child(ExpectServerHello(version=(3, 3)))
+    node = node.add_child(ExpectCertificate())
+    node = node.add_child(ExpectServerHelloDone())
+    # section D.3 of draft 28 states that client that receives TLS 1.2
```
Comment (nmav) [inline scripts/test-tls13-0rtt-garbage.py]: Doesn't this assume that TLS1.4 will not support 1-rtt? Maybe use a much larger version?

**Comment by tomato42** [comment]
@nmav typo fixed

**Comment by tomato42** [comment]
Thanks for review!

**Pull request merged**